test/unit/math/rev/mat/functor/gradient_test.cpp

```diff
@@ -26,6 +31,51 @@ TEST(AgradAutoDiff, gradient) {
   EXPECT_FLOAT_EQ(x(0) * x(0) + 3 * 2 * x(1), grad_fx(1));
 }
 
+// test threaded AD
+TEST(AgradAutoDiff, gradient_threaded) {
+  fun1 f;
+  Matrix<double, Dynamic, 1> x(2);
+  x << 5, 7;
+  double fx_ref;
+  Matrix<double, Dynamic, 1> grad_fx_ref;
+  stan::math::gradient(f, x, fx_ref, grad_fx_ref);
+  EXPECT_FLOAT_EQ(5 * 5 * 7 + 3 * 7 * 7, fx_ref);
+  EXPECT_EQ(2, grad_fx_ref.size());
+  EXPECT_FLOAT_EQ(2 * x(0) * x(1), grad_fx_ref(0));
+  EXPECT_FLOAT_EQ(x(0) * x(0) + 3 * 2 * x(1), grad_fx_ref(1));
+
+  auto thread_job = [&]() {
+    double fx;
+    VectorXd grad_fx;
+    stan::math::gradient(fun1(), x, fx, grad_fx);
+    VectorXd res(1 + grad_fx.size());
+    res(0) = fx;
+    res.tail(grad_fx.size()) = grad_fx;
+    return res;
+  };
+
+  std::vector<std::future<VectorXd>> ad_futures;
+
+  for (std::size_t i = 0; i < 100; i++) {
```

Comment:
I think it'd be good to give the test some opportunities to mess up - right now every thread is running the same function with the exact same inputs and outputs. Can you add some work in there and some dependence on `i` maybe?